### DISCUSSION TITLE
Trying to get trusted header auth to work

### DISCUSSION THREAD

**Discussion by Atlasfreak** [muety/wakapi#644] [Q&A] [answered]
Hi,
I am trying to get trusted header auth to work with authentik as the sso provider, apache with mod_auth_openidc as the proxy and running wakapi via docker compose.
The problem is I do not know why it isn't working. It could be wrong header names or I haven't configured the trusted ips correctly. With the ips I also just do not know which one to use as docker compose assigns the container to a random free network making it impossible to hardcode the gateway ip and idk if "host.docker.internal" works.

Anyone might know where the problem lies or has already managed to get trusted_header auth working with docker compose?

**Comment by muety** [comment]
Hi @Atlasfreak! Getting these three tools properly hooked up together (in Docker) indeed sounds like a non-trivial endeavor. Just a couple of thought from my side that might help you debug this:

First of all, you'll have to set `WAKAPI_TRUSTED_HEADER_AUTH=true` and `WAKAPI_TRUSTED_HEADER_AUTH_KEY=REMOTE_USER` (case-insensitive).

Then, regarding IPs: if you're running Wakapi in Docker, while Apache is running on the host directly, requests relayed from Apache will probably appear to come from the Docker network bridge's gateway address. That's `172.17.0.1` by default, you can double-check using `docker inspect`. If Apache is running in Docker as well, I'd recommend to create a new network (e.g. name it `reverse_proxy_net` and connect both containers to it. The container's IPs will then have IPs like `172.<i>.0.<j>`, where `i` is the index of the network plus one (e.g. `2` if you only have one Docker network configured) and `j` is the index of the container plus one in the order you started them (e.g. `2` for the first container connected to this network). Again, you can use `docker inspect` for investigation.

Another issue might be the value that `mod_auth_openidc` passes on in the trusted auth header field. According to their [docs](https://github.com/OpenIDC/mod_auth_openidc?tab=readme-ov-file#overview), it's set to the concatenation of the token's `sub` and `iss` claims (like `[sub]@[iss]`). Wakapi, however, requires the header to contain nothing but a plain, valid Wakapi user name (e.g. `muety`). Apparently, you can customize (see [here](https://github.com/OpenIDC/mod_auth_openidc/wiki/Authorization)) the header field (using `OIDCRemoteUserClaim`?). So the preferred option here would be to configure a `wakapi_username` (or sth.) attribute for your user in Authentik and then configure the reverse proxy to use the token claim corresponding to that attribute as the value for `REMOTE_USER`. If that's not an option, we could think of extending Wakapi in a way that not only usernames, but also e-mail addresses are accepted for trusted header authentication.

In general, a tool that might help you debug HTTP-related problems like this is https://rbaskets.in/web, I'm using it regularly.

Hope that helps. Keep us posted.

**Reply by Atlasfreak** [reply]
Thanks for the in depth help.
I am pretty sure that the problem is the IP as docker compose always assigns a random network with a random gateway so I cannot hardcode it. In my case it is most often placed in the 192.168.x.x range and the Gateway is also always somewhere in that range but it is different everytime I reboot the container. So harcoding this is not possible.
I sadly do not have enough knowledge of docker to know how to either fix the network where wakapi is placed or if there are any variables which can be used in a compose file to get the gateway address.

**Reply by Atlasfreak** [reply, marked as answer]
Finally got it to work.
I set the `network_mode` variable to `bridge` so that the wakapi container would be attached to the default docker network.
Then I could set the `WAKAPI_TRUST_REVERSE_PROXY_IPS` to `172.17.0.1`. Additionally I needed to set `OIDCAuthNHeader Remote-User` in the apache config.
Also you can change the value `mod_auth_openidc` passes to wakapi with `OIDCRemoteUserClaim`. I set this to `sub` and adjusted the wakapi username to fit the username in authentik.
Thanks again for the help @muety rbaskets did really help.
Apparently Apache never sent any Remote-User Header at all not even REMOTE_USER

**Comment by blesswinsamuel** [comment]
I use kubernetes with Traefik as the reverse proxy, which always gets a variable IP address. I ended up listing all the possible IPs. 😭 

```
WAKAPI_TRUST_REVERSE_PROXY_IPS=10.42.0.0,10.42.3.0,10.42.0.1,10.42.3.1,10.42.0.2,10.42.3.2,10.42.0.3,10.42.3.3,10.42.0.4,10.42.3.4,10.42.0.5,10.42.3.5,10.42.0.6,10.42.3.6,10.42.0.7,10.42.3.7,10.42.0.8,10.42.3.8,10.42.0.9,10.42.3.9,10.42.0.10,10.42.3.10,10.42.0.11,10.42.3.11,10.42.0.12,10.42.3.12,10.42.0.13,10.42.3.13,10.42.0.14,10.42.3.14,10.42.0.15,10.42.3.15,10.42.0.16,10.42.3.16,10.42.0.17,10.42.3.17,10.42.0.18,10.42.3.18,10.42.0.19,10.42.3.19,10.42.0.20,10.42.3.20,10.42.0.21,10.42.3.21,10.42.0.22,10.42.3.22,10.42.0.23,10.42.3.23,10.42.0.24,10.42.3.24,10.42.0.25,10.42.3.25,10.42.0.26,10.42.3.26,10.42.0.27,10.42.3.27,10.42.0.28,10.42.3.28,10.42.0.29,10.42.3.29,10.42.0.30,10.42.3.30,10.42.0.31,10.42.3.31,10.42.0.32,10.42.3.32,10.42.0.33,10.42.3.33,10.42.0.34,10.42.3.34,10.42.0.35,10.42.3.35,10.42.0.36,10.42.3.36,10.42.0.37,10.42.3.37,10.42.0.38,10.42.3.38,10.42.0.39,10.42.3.39,10.42.0.40,10.42.3.40,10.42.0.41,10.42.3.41,10.42.0.42,10.42.3.42,10.42.0.43,10.42.3.43,10.42.0.44,10.42.3.44,10.42.0.45,10.42.3.45,10.42.0.46,10.42.3.46,10.42.0.47,10.42.3.47,10.42.0.48,10.42.3.48,10.42.0.49,10.42.3.49,10.42.0.50,10.42.3.50,10.42.0.51,10.42.3.51,10.42.0.52,10.42.3.52,10.42.0.53,10.42.3.53,10.42.0.54,10.42.3.54,10.42.0.55,10.42.3.55,10.42.0.56,10.42.3.56,10.42.0.57,10.42.3.57,10.42.0.58,10.42.3.58,10.42.0.59,10.42.3.59,10.42.0.60,10.42.3.60,10.42.0.61,10.42.3.61,10.42.0.62,10.42.3.62,10.42.0.63,10.42.3.63,10.42.0.64,10.42.3.64,10.42.0.65,10.42.3.65,10.42.0.66,10.42.3.66,10.42.0.67,10.42.3.67,10.42.0.68,10.42.3.68,10.42.0.69,10.42.3.69,10.42.0.70,10.42.3.70,10.42.0.71,10.42.3.71,10.42.0.72,10.42.3.72,10.42.0.73,10.42.3.73,10.42.0.74,10.42.3.74,10.42.0.75,10.42.3.75,10.42.0.76,10.42.3.76,10.42.0.77,10.42.3.77,10.42.0.78,10.42.3.78,10.42.0.79,10.42.3.79,10.42.0.80,10.42.3.80,10.42.0.81,10.42.3.81,10.42.0.82,10.42.3.82,10.42.0.83,10.42.3.83,10.42.0.84,10.42.3.84,10.42.0.85,10.42.3.85,10.42.0.86,10.42.3.86,10.42.0.87,10.42.3.87,10.42.0.88,10.42.3.88,10.42.0.89,10.42.3.89,10.42.0.90,10.42.3.90,10.42.0.91,10.42.3.91,10.42.0.92,10.42.3.92,10.42.0.93,10.42.3.93,10.42.0.94,10.42.3.94,10.42.0.95,10.42.3.95,10.42.0.96,10.42.3.96,10.42.0.97,10.42.3.97,10.42.0.98,10.42.3.98,10.42.0.99,10.42.3.99,10.42.0.100,10.42.3.100,10.42.0.101,10.42.3.101,10.42.0.102,10.42.3.102,10.42.0.103,10.42.3.103,10.42.0.104,10.42.3.104,10.42.0.105,10.42.3.105,10.42.0.106,10.42.3.106,10.42.0.107,10.42.3.107,10.42.0.108,10.42.3.108,10.42.0.109,10.42.3.109,10.42.0.110,10.42.3.110,10.42.0.111,10.42.3.111,10.42.0.112,10.42.3.112,10.42.0.113,10.42.3.113,10.42.0.114,10.42.3.114,10.42.0.115,10.42.3.115,10.42.0.116,10.42.3.116,10.42.0.117,10.42.3.117,10.42.0.118,10.42.3.118,10.42.0.119,10.42.3.119,10.42.0.120,10.42.3.120,10.42.0.121,10.42.3.121,10.42.0.122,10.42.3.122,10.42.0.123,10.42.3.123,10.42.0.124,10.42.3.124,10.42.0.125,10.42.3.125,10.42.0.126,10.42.3.126,10.42.0.127,10.42.3.127,10.42.0.128,10.42.3.128,10.42.0.129,10.42.3.129,10.42.0.130,10.42.3.130,10.42.0.131,10.42.3.131,10.42.0.132,10.42.3.132,10.42.0.133,10.42.3.133,10.42.0.134,10.42.3.134,10.42.0.135,10.42.3.135,10.42.0.136,10.42.3.136,10.42.0.137,10.42.3.137,10.42.0.138,10.42.3.138,10.42.0.139,10.42.3.139,10.42.0.140,10.42.3.140,10.42.0.141,10.42.3.141,10.42.0.142,10.42.3.142,10.42.0.143,10.42.3.143,10.42.0.144,10.42.3.144,10.42.0.145,10.42.3.145,10.42.0.146,10.42.3.146,10.42.0.147,10.42.3.147,10.42.0.148,10.42.3.148,10.42.0.149,10.42.3.149,10.42.0.150,10.42.3.150,10.42.0.151,10.42.3.151,10.42.0.152,10.42.3.152,10.42.0.153,10.42.3.153,10.42.0.154,10.42.3.154,10.42.0.155,10.42.3.155,10.42.0.156,10.42.3.156,10.42.0.157,10.42.3.157,10.42.0.158,10.42.3.158,10.42.0.159,10.42.3.159,10.42.0.160,10.42.3.160,10.42.0.161,10.42.3.161,10.42.0.162,10.42.3.162,10.42.0.163,10.42.3.163,10.42.0.164,10.42.3.164,10.42.0.165,10.42.3.165,10.42.0.166,10.42.3.166,10.42.0.167,10.42.3.167,10.42.0.168,10.42.3.168,10.42.0.169,10.42.3.169,10.42.0.170,10.42.3.170,10.42.0.171,10.42.3.171,10.42.0.172,10.42.3.172,10.42.0.173,10.42.3.173,10.42.0.174,10.42.3.174,10.42.0.175,10.42.3.175,10.42.0.176,10.42.3.176,10.42.0.177,10.42.3.177,10.42.0.178,10.42.3.178,10.42.0.179,10.42.3.179,10.42.0.180,10.42.3.180,10.42.0.181,10.42.3.181,10.42.0.182,10.42.3.182,10.42.0.183,10.42.3.183,10.42.0.184,10.42.3.184,10.42.0.185,10.42.3.185,10.42.0.186,10.42.3.186,10.42.0.187,10.42.3.187,10.42.0.188,10.42.3.188,10.42.0.189,10.42.3.189,10.42.0.190,10.42.3.190,10.42.0.191,10.42.3.191,10.42.0.192,10.42.3.192,10.42.0.193,10.42.3.193,10.42.0.194,10.42.3.194,10.42.0.195,10.42.3.195,10.42.0.196,10.42.3.196,10.42.0.197,10.42.3.197,10.42.0.198,10.42.3.198,10.42.0.199,10.42.3.199,10.42.0.200,10.42.3.200,10.42.0.201,10.42.3.201,10.42.0.202,10.42.3.202,10.42.0.203,10.42.3.203,10.42.0.204,10.42.3.204,10.42.0.205,10.42.3.205,10.42.0.206,10.42.3.206,10.42.0.207,10.42.3.207,10.42.0.208,10.42.3.208,10.42.0.209,10.42.3.209,10.42.0.210,10.42.3.210,10.42.0.211,10.42.3.211,10.42.0.212,10.42.3.212,10.42.0.213,10.42.3.213,10.42.0.214,10.42.3.214,10.42.0.215,10.42.3.215,10.42.0.216,10.42.3.216,10.42.0.217,10.42.3.217,10.42.0.218,10.42.3.218,10.42.0.219,10.42.3.219,10.42.0.220,10.42.3.220,10.42.0.221,10.42.3.221,10.42.0.222,10.42.3.222,10.42.0.223,10.42.3.223,10.42.0.224,10.42.3.224,10.42.0.225,10.42.3.225,10.42.0.226,10.42.3.226,10.42.0.227,10.42.3.227,10.42.0.228,10.42.3.228,10.42.0.229,10.42.3.229,10.42.0.230,10.42.3.230,10.42.0.231,10.42.3.231,10.42.0.232,10.42.3.232,10.42.0.233,10.42.3.233,10.42.0.234,10.42.3.234,10.42.0.235,10.42.3.235,10.42.0.236,10.42.3.236,10.42.0.237,10.42.3.237,10.42.0.238,10.42.3.238,10.42.0.239,10.42.3.239,10.42.0.240,10.42.3.240,10.42.0.241,10.42.3.241,10.42.0.242,10.42.3.242,10.42.0.243,10.42.3.243,10.42.0.244,10.42.3.244,10.42.0.245,10.42.3.245,10.42.0.246,10.42.3.246,10.42.0.247,10.42.3.247,10.42.0.248,10.42.3.248,10.42.0.249,10.42.3.249,10.42.0.250,10.42.3.250,10.42.0.251,10.42.3.251,10.42.0.252,10.42.3.252,10.42.0.253,10.42.3.253,10.42.0.254,10.42.3.254
```

Would be great if there was a way to configure subnets in `WAKAPI_TRUST_REVERSE_PROXY_IPS`. Example: `10.42.0.0/16`.

**Reply by muety** [reply]
Good point! Added this as a feature request, #665.
